CONTRIBUTING.md

```diff
@@ -0,0 +1,203 @@
+``` 
+
+### Constants
+
+If a constant is used in more than one file, define it in `Common.h`. For
+constants with long names, use underscores to separate words in the constant
+name. Use all caps (screaming snake case).
+
+```c++
+   constexpr double Pi = 3.1415; // No, the constant name should be all caps
+   constexpr double SQRT_TWO = 1.4142 // Yes
+   constexpr double SQRTTWO_ = 1.4142 // No, there is a trailing underscore
+   constexpr double EXP = 2.7183 // Yes   
+```
+
+### Pointers and references
+
+The pointer `*` or reference `&` belong to the type and there should be no space between them and the type name.
+```c++
+double* x;     // Yes
+int& n;        // Yes
+double *x, *y; // No, the pointer symbol is a part of `double*` type
+int & n;       // No, the reference symbol is a part of `int&` type
+```
+
+### Indentation
+Use only spaces for indentation, not tabs. Indent size is 4 spaces.
+
+When defining a class, the code blocks after `private`, `public` and `protected` should be aligned with opening/closing braces. There should be an empty line before each definition (except the first one). See example below.
+```c++
+class SomeClass
+{
+public:
+  SomeClass();
+  ~SomeClass();
+
+private:
+  int some_variable_;
+
+protected:
+  void someFunction();
+};
+```
+
+### Braces
+All braces should follow Allman style:
+```c++
+namespace someNamespace
+{
+  //some code 
+}
+```
+For short functions (i.e., empty constructor), do not inline braces.
+```c++
+ClassA::ClassA()
+{
+}
+```
+Have opening brace at the next line following  `for`, `if`, or `while` statement. When using `else`, follow the example below. 
+```c++
+if (cond == true)
+{
+  // some code
+}
+else
+{
+  // some other code
+}
+ ```
+Have a space between keywords `for`, `while` and `if` and the parenthesis as shown here:
+```c++
+for (int i = 0; i < n; ++i)
+{
+  // some code
+} 
+```
+
+Do not use one-line `if`s and `for`s. Always use braces.
+
+### Use of spaces and newlines
+There should be spaces between arithmetic operators. 
+```c++
+x = c * (a + b);  //Yes
+x = c*(a+b).      // No, the clarity is better if there are spaces between binary operators and operands.
+```
+When defining member functions, use one empty line between the functions.
+```c++
+struct MyStruct
+{
+  int memberFunction()
+  {
+    // some code
+  }
+
+  int anotherMemberFunction()
+  {
+    // some other code
+  }
+};
+```
+Leave one empty line between all the includes and the first line of the actual code. 
+```c++
+#include <iostream>
+
+int main()
+{
+  std::cout 
+}
+```
+
+Also, leave one empty line between `system` includes and `GridKit` includes, i.e.,
+```c++
+#include <cstring>
+
+#include <SparseMatrix/CooMatrix.hpp>
+
+int main()
+{
+  //some code
+  return 0;
+}
+```
+The `system` includes should always be listed first.
+
+### Using namespaces
+All classes should be in namespace `GridKit`. If needed, define additional namespaces inside `GridKit`.
```

Comment:
@pelesh I've noticed there are some namespaces, such as [ModelLib](https://github.com/ORNL/GridKit/blob/60647ed1053a2ed3dab0c45855801246e6d0ee01/ModelEvaluatorImpl.hpp#L66) that do not follow this trend. Is this meant to be an exception or should this be fixed?

Reply:
This should be fixed. However, I suggest we grandfather noncompliant code and merge developer guidelines first. Then we can document issues where we don't see compliance with the guidelines.